File: ingest/startIngest.r
# Call with
#
# irule -F startIngest.r "*user='dlinssen'" "*token='creepy-click'"

startIngest {
    *srcColl = "/nlmumc/ingest/zones/*token";

    *hasDropZonepermission = "";
    checkDropZoneACL(*user, *hasDropZonepermission);
    if (*hasDropZonepermission == "false") {
        # -818000 CAT_NO_ACCESS_PERMISSION
        failmsg(-818000, "User '*user' has insufficient DropZone permissions on /nlmumc/ingest/zones");
    }

    if (errorcode(msiObjStat(*srcColl,*out)) < 0) {
        # CAT_UNKNOWN_COLLECTION
        failmsg(-814000, "Unknown ingest zone *token");
    }

    *state = ""; *project = ""; *title = "";
    foreach (*av in SELECT META_COLL_ATTR_NAME, META_COLL_ATTR_VALUE WHERE COLL_NAME == *srcColl) {
        if ( *av.META_COLL_ATTR_NAME == "project" ) {
            *project = *av.META_COLL_ATTR_VALUE;
        }
        if ( *av.META_COLL_ATTR_NAME == "title" ) {
            *title = *av.META_COLL_ATTR_VALUE;
        }
        if ( *av.META_COLL_ATTR_NAME == "state" ) {
            *state = *av.META_COLL_ATTR_VALUE;
        }
    }

    # Check for project's ingest resource status to start ingestion
    getCollectionAVU("/nlmumc/projects/*project","resource",*resource,"","true");
    *rescStatus = ""
    foreach (*r in select RESC_STATUS where RESC_NAME = *resource ) {
        *rescStatus = *r.RESC_STATUS;
    }

    # Only down is a real status. Anything but "down" means up
    if ( *rescStatus == "down" ) {
        # -831000 CAT_INVALID_RESOURCE
        failmsg(-831000, "Ingest disabled for this resource.");
    }

    # Check for valid state to start ingestion
    if ( *state != "open" && *state != "warning-validation-incorrect" ) {
        failmsg(-1, "Invalid state to start ingestion.");
    }

    if ( *project == "" ) {
        failmsg(-1, "Project is empty.");
    }

    msiWriteRodsLog("Setting status *srcColl to in queue", 0);
    msiAddKeyVal(*stateKV, "state", "in-queue-for-validation");
    msiSetKeyValuePairsToObj(*stateKV, *srcColl, "-C");

    # Remove validateState. Required due to parallel execution of irods queue
    getCollectionAVU(*srcColl,"validateState",*validateState,"","false");
    if ( *validateState != "" ) {
        msiAddKeyVal(*validateStateKV, "validateState", *validateState );
        msiRemoveKeyValuePairsFromObj(*validateStateKV, *srcColl, "-C");
    }
    msiWriteRodsLog("Starting validation of *srcColl", 0)
    *validationResult = ""
    validate_metadata(*srcColl, *validationResult)
    if (*validationResult == "true") {
        msiWriteRodsLog("Validation result is OK, starting ingest of *srcColl", 0)
        # On a new delay queue, as we do not want to repeat this part after failure as above
        # We also do not want any repeats of this, as this would create a new project collection
        delay("<PLUSET>1s</PLUSET><EF>30s REPEAT 0 TIMES</EF>") {
            performIngest(*srcColl, *project, *title, *user, *token);
        }
    }
    else{
        setErrorAVU(*srcColl, "state", "warning-validation-incorrect", "Metadata is incorrect")
    }
}

INPUT *user="",*token=""
OUTPUT ruleExecOut

File: python/start_ingest.py
@make(inputs=[0, 1], outputs=[], handler=Output.STORE)
def start_ingest(ctx, username, token):
    """
    Start an ingest
       Irods pre-ingest checks
       Metadata pre-ingest checks
        If those went well, call perform ingest

    Parameters
    ----------
    ctx : Context
        Combined type of callback and rei struct.
    username: str
        The username, ie 'dlinssen'
    token: str
        The token, ie 'crazy-frog'
    """
    source_collection = "/nlmumc/ingest/zones/{}".format(token)

    # Check if ingesting user has dropzone permissions
    has_dropzone_permission = ctx.callback.checkDropZoneACL(username, "")["arguments"][1]
    if has_dropzone_permission != "true":
        ctx.callback.msiExit(
            "-818000", "User '{}' has insufficient DropZone permissions on /nlmumc/ingest/zones".format(username)
        )

    # Check if dropzone exists
    try:
        ctx.callback.msiObjStat(source_collection, irods_types.RodsObjStat())
    except RuntimeError:
        # -814000 CAT_UNKNOWN_COLLECTION
        ctx.callback.msiExit("-814000", "Unknown ingest zone")

    # Get dropzone metadata
    project_id = ctx.callback.getCollectionAVU(source_collection, "project", "", "", "true")["arguments"][2]

    # Check if project path exists
    try:
        ctx.callback.msiObjStat("/nlmumc/projects/{}".format(project_id), irods_types.RodsObjStat())
    except RuntimeError:
        # -814000 CAT_UNKNOWN_COLLECTION
        ctx.callback.msiExit("-814000", "Unknown project: {}".format(project_id))

    title = ctx.callback.getCollectionAVU(source_collection, "title", "", "", "true")["arguments"][2]
    state = ctx.callback.getCollectionAVU(source_collection, "state", "", "", "true")["arguments"][2]

    # Get resource availability -- check ingest & destination resource
    available = ctx.callback.get_project_resource_availability(project_id, "true", "true", "false", "")["arguments"][4]
    # Project or ingest resource is not available, abort ingest
    if available != "true":
        # -831000 CAT_INVALID_RESOURCE
        ctx.callback.msiExit(
            "-831000", "The project or ingest resource is disabled for this project '{}'".format(project_id)
        )

    # Check for valid state to start ingestion
    if state != "open" and state != "warning-validation-incorrect":
        ctx.callback.msiExit("-1", "Invalid state to start ingestion.")

    ctx.callback.msiWriteRodsLog("Starting validation of {}:".format(source_collection), 0)
    # Set 'state' AVU to 'validating'
    ctx.callback.setCollectionAVU(source_collection, "state", "validating")

    validation_result = ctx.callback.validate_metadata(source_collection, "")["arguments"][1]
    if validation_result == "true":
        ctx.callback.msiWriteRodsLog(
            "Validation result OK {}. Setting status to 'in-queue-for-ingestion'".format(source_collection), 0
        )
        ctx.callback.setCollectionAVU(source_collection, "state", "in-queue-for-ingestion")
        ctx.delayExec(
            "<PLUSET>1s</PLUSET><EF>30s REPEAT 0 TIMES</EF>",
            "perform_ingest('{}', '{}', '{}', '{}')".format(project_id, title, username, token),
            "",
        )
    else:
        ctx.callback.setErrorAVU(source_collection, "state", "warning-validation-incorrect", "Metadata is incorrect")
